File: cv_12.py
from numpy import *
from pyproj import *

def loadPoints(file_text):
    #Load points
    uv = []
    with open(file_text, 'r') as file:
        for line in file:
            ut, vt = line.strip().split()
            u, v = float(ut), float(vt)
            uv.append([u, v])
    return array(uv)

def samplePoints(umin, umax, vmin, vmax, Du, Dv):
    #Sample grid points
    uv = []
    with open("points.txt", "w") as file:
        for u in range(umin,umax,Du):
            for v in range(vmin,vmax,Dv):
                uv.append([v, u])
    return array(uv)

def sampleMeridians(umin, umax, vmin, vmax, Dv, du):
    #Sample meridians
    uv = []
    for v in arange(vmin, vmax, Dv):
        for u in arange(umin, umax, du):
            uv.append([v, u])
    return array(uv)

def sampleParallels(umin, umax, vmin, vmax, Du, dv):
    #Sample parallels
    uv = []
    for u in arange(umin, umax, Du):
        for v in arange(vmin, vmax, dv):
            uv.append([v, u])
    return array(uv)

def writePoints(file_text, u, v, x, y, a = [], b = []):
    #Write points to file
    with open (file_text, "w") as file:
        for i in range(len(u)):
            #Create line
            line = str(u[i]) + '\t' + str(v[i]) + '\t' + str(x[i]) + '\t' + str(y[i]) + '\t'

            #Tissote indicatrix parameters computed
            if len(a) > 0:
                line += str(a[i]) + '\t' + str(b[i])

            #Terminate line
            line += '\n'

            #Write line
            file.write(line)


#Define extent and steps
umin = -80
umax = 81
vmin = -180
vmax = 181
Du = 10
Dv = 10
dv = Dv/10
du = Du/10

#Sample points, meridians and parallels
uv = samplePoints(umin, umax, vmin, vmax, Du, Dv)
uvm = sampleMeridians(umin, umax, vmin, vmax, Dv, du)
uvp = sampleParallels(umin, umax, vmin, vmax, Dv, du)

#Load Europe
uve = loadPoints('continents\eur.txt')

#Define projection
sinu = Proj(proj='sinu', R=6380)

#Compute coordinates
[x,y] = sinu(uv[:, 0], uv[:, 1])
[xm,ym] = sinu(uvm[:, 0], uvm[:, 1])
[xp,yp] = sinu(uvp[:, 0], uvp[:, 1])
[xe,ye] = sinu(uve[:, 1], uve[:, 0])

#Compute distortions [a,b] for grid points
res = sinu.get_factors(uv[:, 0], uv[:, 1])
a,b = res.tissot_semimajor, res.tissot_semiminor

#Write points to files
writePoints("points_sinu.txt", uv[:, 1], uv[:, 0], x, y, a, b)
writePoints("meridians_sinu.txt", uvm[:, 1], uvm[:, 0], xm, ym, [], [])
writePoints("parallels_sinu.txt", uvp[:, 1], uvp[:, 0], xp, yp, [], [])
writePoints("europe_sinu.txt", uve[:, 1], uve[:, 0], xe, ye, [], [])

File: cv_12.m
clc
clear

format long g
axis equal

% Define extent
umin = -80;
umax = 80;
vmin = -180;
vmax = 180;
Du = 10;
Dv = 10;
dv = Dv/10;
du = Du/10;

R = 6380;

%Load graticule points
P = load("points_sinu.txt");
u = P(:, 1);
x = P(:, 3);
y = P(:, 4);
a = P(:, 5);
b = P(:, 6);

%Create W matrix
W = cos(u*pi/180);

%Airy + complex local
h2a = ((a-1).^2+(b-1).^2)/2;
h2c = (abs(a-1) + abs(b-1))/2 + a./b - 1;

%Airy global, non-weighted
H2A = mean(h2a(:));
H2C = mean(h2c(:));
HA = sqrt(H2A);
HC = sqrt(H2C);

%Airy global, weighted
H2AN = W.*h2a;
H2AW = sum(H2AN(:))/sum(W(:));
HAW = sqrt(H2AW);

H2CN = W.*h2c;
H2CW = sum(H2CN(:))/sum(W(:));
HCW = sqrt(H2CW);

%Load meridians
M = load('meridians_sinu.txt');
xm = M(:, 3);
ym = M(:, 4);

%Load parallels
P = load('parallels_sinu.txt');
xp = P(:, 3);
yp = P(:, 4);

%Load contients
EU = load('europe_sinu.txt');
xeu = EU(:, 3);
yeu = EU(:, 4);

%Draw meridians
hold on
nm = length(umin:du:umax)
for i = 1:nm:length(xm)
    plot(xm(i:i+nm - 1), ym(i:i+nm - 1), 'k');
end

%Draw parallels
np = length(vmin:dv:vmax)
for i = 1:np:length(xp)
    plot(xp(i:i+np - 1), yp(i:i+np - 1), 'k');
end
    
%Draw continents
plot(xeu, yeu, 'b');

%Map scale
M = 100000000;
Muv = M./a;

%Convert [x,y,M] to meshgrid
xg = linspace(min(x), max(x), 100);
yg = linspace(min(y), max(y), 100);
[X,Y] = meshgrid(xg, yg);
Z = griddata(x, y, Muv, X, Y, 'cubic');

%Draw contour lines
[C, h] = contour(X, Y, Z, [20000000:10000000:300000000], 'LineColor', 'r', 'LineWidth' ,2);
clabel(C, h, 'Color', 'r', 'labelspacing', 1000)